### ISSUE TITLE
Running nn.GroupNorm on CPU and GPU return different results

### ISSUE THREAD

**Issue by cckao** [pytorch/pytorch#54293]
## 🐛 Bug

Running nn.GroupNorm on CPU and GPU return different results

## To Reproduce

```python
import torch
import torch.nn as nn

cuda = 'cuda:0'
with torch.no_grad():
    cpu = torch.rand(1, 32, 256, 256, 72, device='cpu')
    gpu = cpu.to(cuda)
    m = nn.GroupNorm(4, 32).eval()
    cpu = m.cpu()(cpu)
    gpu = m.to(cuda)(gpu)
    print(torch.abs(cpu - gpu.cpu()).max())  # 0.0121 on my machine!!
```

## Expected behavior

`torch.abs(cpu - gpu.cpu()).max()` should returns a very small value but I got 0.0121 on my machine.

## Environment

PyTorch version: 1.8.0
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce RTX 2070
Nvidia driver version: 450.102.04
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.8.0
[pip3] torchelastic==0.2.2
[pip3] torchtext==0.9.0
[pip3] torchvision==0.9.0
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.1.74              h6bb024c_0    nvidia
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2020.2                      256  
[conda] mkl-service               2.3.0            py38he904b0f_0  
[conda] mkl_fft                   1.3.0            py38h54f3939_0  
[conda] mkl_random                1.1.1            py38h0573a6f_0  
[conda] numpy                     1.19.2           py38h54aff64_0  
[conda] numpy-base                1.19.2           py38hfa32c7d_0  
[conda] pytorch                   1.8.0           py3.8_cuda11.1_cudnn8.0.5_0    pytorch
[conda] torchelastic              0.2.2                    pypi_0    pypi
[conda] torchtext                 0.9.0                      py38    pytorch
[conda] torchvision               0.9.0                py38_cu111    pytorch

## Additional context

I run it on docker image `pytorch/pytorch:1.8.0-cuda11.1-cudnn8-runtime`.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @ngimel

**Comment by zou3519** [comment]
Tentatively marking high-pri for silent correctness issue. I can reproduce this with the 1.8 binary.

This might be a numerical stability issue. The elements with the largest absolute values on my machine are ~1.7 so a difference of 0.0121 is a lot. However there are probably a lot of floating-point operations going on beneath the hood. Switching to using float64 makes this problem go away entirely.

**Comment by zou3519** [comment]
At the triage review meeting, we decided that this is expected behavior. We don't accumulate into a higher precision and because there are a very large number of elements being summed together in the GroupNorm it loses precision. The workaround for this is to switch to using float64 for this operation.

**Comment by zou3519** [comment]
Reopening due to additional information. We should confirm the problem is really floating point error

**Comment by ezyang** [comment]
@ngimel says that the implementation has changed between 1.4 and master. cc @xiaomengy

**Comment by xiaomengy** [comment]
This issue is mainly from the numerical stability of computing variance on CPU. There are two issues on CPU.

1. Currently in our GroupNorm implementation, we are using the 1-pass algorithm to compute mean and var of the input. The input of this case is from U[0, 1), so the var will be small compare to the size of the input. In this case, the final subtraction of 1-pass algorithm may suffer from the issue of [cancelation](https://en.wikipedia.org/wiki/Loss_of_significance). To fix this I tried to use Welford algorithm and then the result for the test case above can be reduced from 0.0121 to 0.0030. There is some improvement but it is not as large as we expected. Then we have to go to the 2nd issue.

2. The reduction dim of this test case is 256 * 256 * 72 * 8 which is about 3.8e7. While float32 only has 7 significant figures. When all the input numbers are in the range of [0, 1), the summation during the computation of mean and variance will lead to large numerical errors when we just sum things one by one, which is the CPU case. While on GPU, in order to utilize the GPU threads, we split the whole segment into several sub-segments and sum them separately. Each sub-segment is much smaller. And when we combine the results of each threads in a GPU block, it works as [pairwise summation ](https://en.wikipedia.org/wiki/Pairwise_summation) which has a much better numerical stability. So that's why GPU's result is consistent. In order to fix this, one way we can do is to use [Kahan Summation](https://en.wikipedia.org/wiki/Kahan_summation_algorithm). When applying Kahan Summation as well as Welford algorithm, the result of this test case can be reduce to 7.03e-6.

I prepared https://github.com/pytorch/pytorch/pull/54921 for these changes. Please note there will be some performance loss. I benchmarked the performance for input_size = [256, 1024, 28, 28], group = 8 case. The original average running time on my devgpu is about 550ms, after applying Welford algorithm, it became about 594ms, and after applying Kahan Summation, it became about 640ms. Consider the GroupNorm should not consume the major time of a model, I think applying both of them should be acceptable. Another way we can do is applying Welford algorithm by default, and add an argument to GroupNorm which determines to use Kahan Summation or not. This is because when the reduction dims are not very large, the gain from Kahan Summation will be limited.

**Comment by ngimel** [comment]
I think not turning Kahan on by default (or maybe not turning it on at all) should be acceptable. Did you benchmark gpu performance also? Kahan is not needed for gpu, so it shouldn't affect GPU perf, right?

**Comment by xiaomengy** [comment]
Yes, in my PR I only use Kahan summation for CPU side. So I didn't benchmark GPU performance. The only concern is for the test case in this issue, as I mentioned, as a special case that the reduction dim was so large and all of the elements are form U[0, 1), applying Welford can only reduce the abs diff from 0.0121 to 0.0030. This risk may happen in the 3D models and applying GroupNorm after activation such as relu. But personally I agree that let's only apply welford 1st, if we enable kahan summation here, we may need to evaluate other add cases.

**Comment by xiaomengy** [comment]
I have updated https://github.com/pytorch/pytorch/pull/54921 to use cascade sum, currently it has similar numerical stability as kahan summation but better performance.

**Comment by xiaomengy** [comment]
https://github.com/pytorch/pytorch/pull/54921 should fix this.
